torch/optim/lars.py

```diff
@@ -0,0 +1,168 @@
+from .optimizer import Optimizer, required, _use_grad_for_differentiable  # type: ignore[attr-defined]
+import torch
+from typing import List, Optional
+from torch import Tensor
+
+__all__ = ["LARS", "lars"]
+
+
+class LARS(Optimizer):
+    """Implements LARS algorithm."""
+
+    def __init__(
+        self,
+        params,
+        lr=required,
+        weight_decay: float = 0,
+        momentum: float = 0,
+        dampening: float = 0,
+        nesterov: bool = False,
+        trust_coefficient: float = 0.001,
+        eps: float = 1e-8,
+        maximize: bool = False,
+        differentiable: bool = False,
+    ):
+        if lr is not required and lr < 0.0:
+            raise ValueError(f"Invalid learning rate: {lr}")
+        if weight_decay < 0.0:
+            raise ValueError(f"Invalid weight decay value: {weight_decay}")
+        if momentum < 0.0:
+            raise ValueError(f"Invalid momentum value: {momentum}")
+        if nesterov and (momentum <= 0 or dampening != 0):
+            raise ValueError("Nesterov momentum requires a momentum and zero dampening")
+
+        defaults = dict(
+            lr=lr,
+            weight_decay=weight_decay,
+            momentum=momentum,
+            dampening=dampening,
+            nesterov=nesterov,
+            maximize=maximize,
+            differentiable=differentiable,
+        )
+
+        super().__init__(params, defaults)
+        self.trust_coefficient = trust_coefficient
+        self.eps = eps
+
+    def __setstate__(self, state):
+        super().__setstate__(state)
+        for group in self.param_groups:
+            group.setdefault("nesterov", False)
+            group.setdefault("maximize", False)
+            group.setdefault("differentiable", False)
+
+    @_use_grad_for_differentiable
+    def step(self, closure=None):
+        loss = None
+        if closure is not None:
+            with torch.enable_grad():
+                loss = closure()
+
+        for group in self.param_groups:
+            params_with_grad = []
+            grads = []
+            momentum_buffer_list = []
+
+            for p in group["params"]:
+                if p.grad is not None:
+                    params_with_grad.append(p)
+                    grads.append(p.grad)
+
+                    state = self.state[p]
+
+                    momentum_buffer_list.append(state.get("momentum_buffer"))
+
+            lars(
+                params_with_grad,
+                grads,
+                momentum_buffer_list,
+                lr=group["lr"],
+                weight_decay=group["weight_decay"],
+                momentum=group["momentum"],
+                dampening=group["dampening"],
+                nesterov=group["nesterov"],
+                eps=self.eps,
+                trust_coefficient=self.trust_coefficient,
```

Comment:
Does it make sense to grab these from the group similar to other params here?

Reply:
Yes!